applications/crossbar/src/modules/cb_allotments.erl

```diff
@@ -0,0 +1,326 @@
+validate(Context) ->
+    validate_allotments(Context, cb_context:req_verb(Context)).
+
+-spec validate(cb_context:context(), path_token()) -> cb_context:context().
+validate(Context, ?CONSUMED) ->
+    validate_consumed(Context, cb_context:req_verb(Context)).
+
+-spec validate_allotments(cb_context:context(), http_method()) -> cb_context:context().
+validate_allotments(Context, ?HTTP_GET) ->
+    load_allotments(Context);
+validate_allotments(Context, ?HTTP_POST) ->
+    cb_context:validate_request_data(<<"allotments">>, Context, fun on_successful_validation/1).
+
+-spec validate_consumed(cb_context:context(), http_method()) -> cb_context:context().
+validate_consumed(Context, ?HTTP_GET) ->
+    load_consumed(Context).
+
+-spec post(cb_context:context()) -> cb_context:context().
+post(Context) ->
+    update_allotments(Context).
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+%%--------------------------------------------------------------------
+%% @private
+%% @doc
+%% Load a document from the database
+%% @end
+%%--------------------------------------------------------------------
+-spec load_allotments(cb_context:context()) -> cb_context:context().
+load_allotments(Context) ->
+    Context1 = maybe_handle_load_failure(crossbar_doc:load(?PVT_TYPE, Context)),
+    Allotments = wh_json:get_json_value(?PVT_ALLOTMENTS, cb_context:doc(Context1), wh_json:new()),
+    cb_context:set_resp_data(Context1, Allotments).
+
+
+-spec load_consumed(cb_context:context()) -> cb_context:context().
+load_consumed(Context) ->
+    Context1 = crossbar_doc:load(?PVT_TYPE, Context),
+    Allotments = wh_json:get_json_value(?PVT_ALLOTMENTS, cb_context:doc(Context1), wh_json:new()),
+    load_consumed(Context, Allotments).
+
+-spec load_consumed(cb_context:context(), wh_json:objects()) -> cb_context:context().
+load_consumed(Context, Allotments) ->
+    ReqCycleStart = cb_context:req_param(Context, <<"start">>),
+    ReqCycleEnd = cb_context:req_param(Context, <<"end">>),
+    load_consumed(Context, Allotments, ReqCycleStart, ReqCycleEnd).
+
+-spec load_consumed(cb_context:context(), wh_json:objects(), api_seconds(), api_seconds()) -> cb_context:context().
+load_consumed(Context, Allotments, 'undefined', 'undefined') ->
+    DateTime = calendar:universal_time(),
+    MoDb = wh_util:format_account_mod_id(cb_context:account_id(Context), calendar:datetime_to_gregorian_seconds(DateTime)),
+    {_, _, Result} = wh_json:foldl(fun foldl_consumed_datetime/3, {DateTime, MoDb, []}, Allotments),
+    cb_context:setters(Context
+                       ,[{fun cb_context:set_resp_status/2, 'success'}
+                         ,{fun cb_context:set_resp_data/2, Result}
+                        ]);
+
+load_consumed(Context, Allotments, ReqStart, ReqEnd) ->
+    case normalize_timestamps({ReqStart, ReqEnd}) of
```

Comment:
I'd like to see validation errors (see `cb_context:add_validation_error/4`) used here for both timestamps. Meaning, if both are invalid, two validation errors should be in the response.